In [1]:
import pandas as pd
import os
from pathlib import Path

In [2]:
root = Path(os.getcwd()).parent.parent

#Change for each one of where your data is. For me in Dataset I have all the three folders
parent_directory = os.path.dirname(root)
dataset_path = os.path.join(root,'Dataset')

ADVOCATE = "advocate"
RATEBEER = "ratebeer"
MATCHED = "matched"

ADVOCATE_PATH = os.path.join(dataset_path,ADVOCATE)
RATEBEER_PATH = os.path.join(dataset_path,RATEBEER)
MATCHED_PATH = os.path.join(dataset_path,MATCHED)

**Understanding the difference between the three datasets**

We use the breweries dataset first as it is smallest to understand how it was build

In [3]:
breweries_advocate = pd.read_csv(os.path.join(ADVOCATE_PATH,'breweries.csv'))
breweries_matched = pd.read_csv(os.path.join(MATCHED_PATH,'breweries.csv'),header=1)
breweries_ratebeer = pd.read_csv(os.path.join(RATEBEER_PATH,'breweries.csv'))

print("Length of the three datasets:\n-advocate:",len(breweries_advocate),"\n-matched:", len(breweries_matched),"\n-ratebeer:", len(breweries_ratebeer))

Length of the three datasets:
-advocate: 16758 
-matched: 8281 
-ratebeer: 24189


In [4]:
breweries_matched.head(10)

id          location                                  name  nbr_beers  \
0  10093  Northern Ireland  Strangford Lough Brewing Company Ltd          5   
1  32848  Northern Ireland                   The Sheelin Brewery          4   
2  40360  Northern Ireland           Walled City Brewing Company          6   
3  40309  Northern Ireland                  Ards Brewing Company          7   
4  41205  Northern Ireland                   Barrahooley Brewery          3   
5  45082  Northern Ireland             Bullhouse Brewing Company          6   
6  19223  Northern Ireland            Clanconnel Brewing Company          6   
7  35512  Northern Ireland                  Clearsky Brewing Co.          2   
8  36583  Northern Ireland             Farmageddon Brewing Co-op         22   
9  40310  Northern Ireland                   Glens of Antrim Ale          4   

    id.1        location.1                     name.1  nbr_beers.1      diff  \
0   4959  Northern Ireland           Strangford Lough            5  0.431275   
1  17616  Northern Ireland                    Sheelin            2  0.526388   
2  24866  Northern Ireland                Walled City            3  0.527852   
3  13538  Northern Ireland           Ards Brewing Co.           13  0.554395   
4  22304  Northern Ireland  Barrahooley Craft Brewery            4  0.602544   
5  27006  Northern Ireland  Bullhouse Brewing Company            6  0.658297   
6  12090  Northern Ireland                 Clanconnel            7  0.598105   
7  18096  Northern Ireland                   Clearsky            3  0.474283   
8  19507  Northern Ireland  Farmageddon Brewing Co-Op           23  0.439740   
9  21530  Northern Ireland            Glens of Antrim            3  0.555796   

        sim  
0  0.889062  
1  0.863596  
2  0.954183  
3  0.896098  
4  0.896205  
5  1.000000  
6  0.939808  
7  0.903298  
8  1.000000  
9  0.933282

Do a special case for a certain value to make sure we understand what is hapening

In [5]:
N = 1000
print(breweries_matched.iloc[N])
print(breweries_matched.iloc[N].id)
print(breweries_matched.iloc[N]["id.1"])

id                              41217
location                       Canada
name           Wheelhouse Brewing Co.
nbr_beers                           1
id.1                            18583
location.1                     Canada
name.1             Wheelhouse Brewing
nbr_beers.1                        12
diff                          0.49904
sim                          0.928055
Name: 1000, dtype: object
41217
18583


In [6]:
breweries_advocate[breweries_advocate.id==breweries_matched.iloc[N].id].head()

id location                    name  nbr_beers
2120  41217   Canada  Wheelhouse Brewing Co.          1

In [7]:
breweries_ratebeer[breweries_ratebeer.id==breweries_matched.iloc[N]["id.1"]].head()

id location                name  nbr_beers
1468  18583   Canada  Wheelhouse Brewing         12

The names differ but those are the same breweries. Let's see what are the beers that differ from it. We will look at the different beers of the beers of this N value.

In [8]:
beers_advocate = pd.read_csv(os.path.join(ADVOCATE_PATH,'beers.csv'))
beers_matched = pd.read_csv(os.path.join(MATCHED_PATH,'beers.csv'),header=1)
beers_ratebeer = pd.read_csv(os.path.join(RATEBEER_PATH,'beers.csv'))

print("Length of the three datasets:\n-advocate:",len(beers_advocate),"\n-matched:", len(beers_matched),"\n-ratebeer:", len(beers_ratebeer))

Length of the three datasets:
-advocate: 280823 
-matched: 45640 
-ratebeer: 442081


In [9]:
print(beers_matched.columns)

Index(['abv', 'avg', 'avg_computed', 'avg_matched_valid_ratings', 'ba_score',
       'beer_id', 'beer_name', 'beer_wout_brewery_name', 'brewery_id',
       'brewery_name', 'bros_score', 'nbr_matched_valid_ratings',
       'nbr_ratings', 'nbr_reviews', 'style', 'zscore', 'abv.1', 'avg.1',
       'avg_computed.1', 'avg_matched_valid_ratings.1', 'beer_id.1',
       'beer_name.1', 'beer_wout_brewery_name.1', 'brewery_id.1',
       'brewery_name.1', 'nbr_matched_valid_ratings.1', 'nbr_ratings.1',
       'overall_score', 'style.1', 'style_score', 'zscore.1', 'diff', 'sim'],
      dtype='object')


In [10]:
beers_matched[beers_matched.brewery_id==breweries_matched.iloc[N].id]

abv   avg  avg_computed  avg_matched_valid_ratings  ba_score  beer_id  \
5005  5.3  3.46          3.46                       3.46       NaN   281481   

              beer_name beer_wout_brewery_name  brewery_id  \
5005  Flagship Pale Ale      Flagship Pale Ale       41217   

                brewery_name  ...  brewery_id.1      brewery_name.1  \
5005  Wheelhouse Brewing Co.  ...         18583  Wheelhouse Brewing   

      nbr_matched_valid_ratings.1  nbr_ratings.1 overall_score  \
5005                            3              3           NaN   

                style.1  style_score  zscore.1      diff  sim  
5005  American Pale Ale          NaN -0.430073  0.776539  1.0  

[1 rows x 33 columns]

In [11]:

print(beers_matched[beers_matched.brewery_id==breweries_matched.iloc[N].id].beer_name)
print(beers_matched[beers_matched.brewery_id==breweries_matched.iloc[N].id]["beer_name.1"])

5005    Flagship Pale Ale
Name: beer_name, dtype: object
5005    Wheelhouse Flagship Pale Ale
Name: beer_name.1, dtype: object


In [12]:
beers_ratebeer[beers_ratebeer.brewery_id==breweries_matched.iloc[N]["id.1"]]

beer_id                                          beer_name  brewery_id  \
22457   314963  Wheelhouse / Three Ranges Yellowhead Imperial Red       18583   
22458   372420       Wheelhouse / Yellow Dog First Mate White IPA       18583   
22459   246707                    Wheelhouse Blacksmith Brown Ale       18583   
22460   448321                  Wheelhouse Divide Rides Shred Red       18583   
22461   246706                       Wheelhouse Flagship Pale Ale       18583   
22462   246705                   Wheelhouse Gillnetter Golden Ale       18583   
22463   329038                           Wheelhouse Kazu Maru IPA       18583   
22464   448322                   Wheelhouse Northern Berry Saison       18583   
22465   372412                   Wheelhouse Scurvy Dog Spruce Tip       18583   
22466   372421                           Wheelhouse Smoked Kolsch       18583   
22467   418412                       Wheelhouse Smokehouse Porter       18583   
22468   497953                             Wheelhouse Vitamin Sea       18583   

             brewery_name                 style  nbr_ratings  overall_score  \
22457  Wheelhouse Brewing  American Strong Ale             0            NaN   
22458  Wheelhouse Brewing  India Pale Ale (IPA)            1            NaN   
22459  Wheelhouse Brewing             Brown Ale            2            NaN   
22460  Wheelhouse Brewing             Amber Ale            1            NaN   
22461  Wheelhouse Brewing     American Pale Ale            3            NaN   
22462  Wheelhouse Brewing  Golden Ale/Blond Ale            3            NaN   
22463  Wheelhouse Brewing  India Pale Ale (IPA)            1            NaN   
22464  Wheelhouse Brewing                Saison            1            NaN   
22465  Wheelhouse Brewing  Spice/Herb/Vegetable            2            NaN   
22466  Wheelhouse Brewing                Smoked            2            NaN   
22467  Wheelhouse Brewing                Porter            2            NaN   
22468  Wheelhouse Brewing     American Pale Ale            0            NaN   

       style_score   avg  abv  avg_computed    zscore  \
22457          NaN   NaN  8.0           NaN       NaN   
22458          NaN  2.92  NaN      2.900000       NaN   
22459          NaN  3.11  NaN      3.450000       NaN   
22460          NaN  2.83  NaN      2.500000       NaN   
22461          NaN  3.04  5.3      3.200000 -0.430073   
22462          NaN  2.81  4.8      2.666667       NaN   
22463          NaN  2.79  NaN      2.300000       NaN   
22464          NaN  2.83  NaN      2.500000       NaN   
22465          NaN  3.05  5.3      3.300000       NaN   
22466          NaN  3.05  5.3      3.300000       NaN   
22467          NaN  3.23  6.3      3.800000       NaN   
22468          NaN   NaN  4.8           NaN       NaN   

       nbr_matched_valid_ratings  avg_matched_valid_ratings  
22457                          0                        NaN  
22458                          0                        NaN  
22459                          0                        NaN  
22460                          0                        NaN  
22461                          3                        3.2  
22462                          0                        NaN  
22463                          0                        NaN  
22464                          0                        NaN  
22465                          0                        NaN  
22466                          0                        NaN  
22467                          0                        NaN  
22468                          0                        NaN

In [13]:
beers_advocate[beers_advocate.brewery_id==breweries_matched.iloc[N]["id"]]

beer_id          beer_name  brewery_id            brewery_name  \
23236   281481  Flagship Pale Ale       41217  Wheelhouse Brewing Co.   

                         style  nbr_ratings  nbr_reviews   avg  ba_score  \
23236  American Pale Ale (APA)            1            1  3.46       NaN   

       bros_score  abv  avg_computed    zscore  nbr_matched_valid_ratings  \
23236         NaN  5.3          3.46 -1.202327                          1   

       avg_matched_valid_ratings  
23236                       3.46

So we see that the matched dataset contains the matched data between the two datasets it took care of issues like brewery name/beer name difference. Now let's see how many breweries we really have. However the matched dataset will tell us how many beers each dataset have, but we must check how many of those beers are matched etc...

In [14]:
breweries_ratebeer_solo = breweries_ratebeer[~breweries_ratebeer.id.isin(breweries_matched['id.1'])]
breweries_advocate_solo = breweries_advocate[~breweries_advocate.id.isin(breweries_matched['id'])]


total_len = len(breweries_advocate_solo) + len(breweries_ratebeer_solo) + len(breweries_matched)
print("Previous of the three datasets:\n-advocate:",len(breweries_advocate),"\n-matched:", len(breweries_matched),"\n-ratebeer:", len(breweries_ratebeer))
print("New length of of the three datasets:\n-advocate:",len(breweries_advocate_solo),"\n-matched:", len(breweries_matched),"\n-ratebeer:", len(breweries_ratebeer_solo))
print("The unique number of breweries is:", total_len)

Previous of the three datasets:
-advocate: 16758 
-matched: 8281 
-ratebeer: 24189
New length of of the three datasets:
-advocate: 8477 
-matched: 8281 
-ratebeer: 15954
The unique number of breweries is: 32712


But 24189 -8281 = 15908 and not 15954. But every match has an id in ratebeer what is the issue?

In [15]:
breweries_ratebeer_solo_test = breweries_ratebeer[~breweries_ratebeer.id.isin(breweries_matched['id.1'])]
test = breweries_ratebeer_solo_test[breweries_ratebeer_solo_test.id.isin(breweries_matched['id.1'])]
test.head()

Empty DataFrame
Columns: [id, location, name, nbr_beers]
Index: []

In [16]:
duplicates = breweries_matched['id.1'].duplicated()
duplicate_rows = breweries_matched[breweries_matched['id.1'].duplicated(keep=False)]
#print(len(duplicate_rows))
duplicate_rows.head()
#print(duplicates.any())

id location                    name  nbr_beers   id.1 location.1  \
219  37180  England          Seven Bro7hers          4  20891    England   
220  45243  England  Seven Bro7hers Brewery          0  20891    England   
447  25939  England        Dartmoor Brewery          3   3480    England   
448  22832  England   Dartmoor Brewery Ltd.          0   3480    England   
614  24935    China      Great Leap Brewery         24  12325      China   

                 name.1  nbr_beers.1      diff       sim  
219      Seven Bro7hers            7  0.528802  1.000000  
220      Seven Bro7hers            7  0.450217  0.972413  
447            Dartmoor           16  0.662249  0.955910  
448            Dartmoor           16  0.388168  0.824178  
614  Great Leap Brewing           53  0.382026  0.946753

Oh! So there are duplicates, how to treat them? Well good question, but at least we now know why there was more data in the solo dataset than expected. Let's check the other dataset.

In [17]:
duplicates = breweries_matched['id'].duplicated()
duplicate_rows = breweries_matched[breweries_matched['id'].duplicated(keep=False)]
#print(len(duplicate_rows))
duplicate_rows.head()

Empty DataFrame
Columns: [id, location, name, nbr_beers, id.1, location.1, name.1, nbr_beers.1, diff, sim]
Index: []

No duplication for the id of advocate. But it means that certain ids of brewery have different ids on advocate whereas rateeer has just one value for those.
We will have to be carefull for this depending on what we want to do with the data. Can we make sure that there are no other duplicates then in each dataset, like a same brewerie with different names/id etc..?

**The question remains what is then the real total number of breweries in the dataset, if we consider that duplicates should be considered as one single brewery?**

In [18]:
duplicated_ids_var = breweries_matched[breweries_matched['id.1'].duplicated(keep=False)]
duplicated_ids = duplicated_ids_var['id.1']

duplicated_tes_max_value_id = duplicated_ids_var.groupby("id.1")["id"].min()
print(len(duplicated_tes_max_value_id))


45


In [19]:
breweries_ratebeer_solo = breweries_ratebeer[~breweries_ratebeer.id.isin(breweries_matched['id.1'])]
breweries_advocate_solo = breweries_advocate[~breweries_advocate.id.isin(breweries_matched['id'])]
breweries_matched_solo = breweries_matched[~((breweries_matched["id.1"].isin(duplicated_ids)) & ~(breweries_matched.id.isin(duplicated_tes_max_value_id)))]#Keep only min value of id of duplicated id set and making sure that it is an id that is duplicated in ratebeer

total_len = len(breweries_advocate_solo) + len(breweries_ratebeer_solo) + len(breweries_matched_solo)
print("Previous of the three datasets:\n-advocate:",len(breweries_advocate),"\n-matched:", len(breweries_matched),"\n-ratebeer:", len(breweries_ratebeer))
print("New length of of the three datasets:\n-advocate:",len(breweries_advocate_solo),"\n-matched:", len(breweries_matched_solo),"\n-ratebeer:", len(breweries_ratebeer_solo))
print("The unique number of breweries is:", total_len)

Previous of the three datasets:
-advocate: 16758 
-matched: 8281 
-ratebeer: 24189
New length of of the three datasets:
-advocate: 8477 
-matched: 8235 
-ratebeer: 15954
The unique number of breweries is: 32666


Okay now we want to see if there can be duplicates in the users, beers and maybe ratings. Let's begin with the users.

**USERS DUPLICATES**

In [44]:
users_matched = pd.read_csv(os.path.join(MATCHED_PATH,'users.csv'),header=1)
users_approx_matched = pd.read_csv(os.path.join(MATCHED_PATH,'users_approx.csv'),header=1)
users_advocate = pd.read_csv(os.path.join(ADVOCATE_PATH,'users.csv'))
users_ratebeer = pd.read_csv(os.path.join(RATEBEER_PATH,'users.csv'))

print(len(users_matched))
print(len(users_approx_matched))
print(len(users_advocate))
print(len(users_ratebeer))


3020
3341
153704
70174


Let us first try and understand the difference between the two datasets.

In [45]:
users_approx_matched_unique = users_approx_matched[~users_approx_matched.user_id.isin(users_matched.user_id)]
users_matched_unique = users_matched[~users_matched.user_id.isin(users_approx_matched.user_id)]

print(len(users_approx_matched_unique))
print(len(users_matched_unique))
print(users_approx_matched_unique.sim.max())


309
0
0.966091783079296


In [46]:
print((users_matched[users_matched.user_id=="magicuenca.1185749"]))
print((users_approx_matched[users_approx_matched.user_id=="magicuenca.1185749"]))

            joined location  nbr_ratings  nbr_reviews             user_id  \
1155  1.483009e+09    Spain            3            0  magicuenca.1185749   

       user_name user_name_lower      joined.1 location.1  nbr_ratings.1  \
1155  MAGICuenca      magicuenca  1.481108e+09      Spain              9   

      user_id.1 user_name.1 user_name_lower.1  
1155     437310  MAGICuenca        magicuenca  
            joined location  nbr_ratings  nbr_reviews             user_id  \
0     1.483009e+09    Spain            3            0  magicuenca.1185749   
1260  1.483009e+09    Spain            3            0  magicuenca.1185749   

       user_name user_name_lower      joined.1 location.1  nbr_ratings.1  \
0     MAGICuenca      magicuenca  1.484046e+09      Spain             89   
1260  MAGICuenca      magicuenca  1.481108e+09      Spain              9   

      user_id.1   user_name.1 user_name_lower.1       sim  
0        442761  MAGICuenca91      magicuenca91  0.904534  
1260     437310

Again there seems to be duplication, but first: approx seems to be the likelihood of a user to be the same. ANd herre in approx there is a duplication of a user because a user in one set is similar to two other users names in the other dataset. Maybe the three of them are the same person maybe not. Certain users are not sure to be similar so it will be up to us to decide how we treat them.

Now let's count the duplicates.

In [49]:
users_matched.sample()

joined                 location  nbr_ratings  nbr_reviews  \
793  1.225537e+09  United States, Oklahoma          247          230   

                 user_id     user_name user_name_lower      joined.1  \
793  webbcreative.262113  webbcreative    webbcreative  1.228734e+09   

                  location.1  nbr_ratings.1  user_id.1   user_name.1  \
793  United States, Oklahoma             98      84287  webbcreative   

    user_name_lower.1  
793      webbcreative

In [ ]:
duplicated_users_in_advocate = users_matched[users_matched.user_id.duplicated(keep=False)]
duplicated_users_in_ratebeer = users_matched[users_matched["user_id.1"].duplicated(keep=False)]
print(len(duplicated_users_in_advocate))
print(len(duplicated_users_in_ratebeer))
#No duplicated and verify in the approx file
duplicated_users_in_advocate = users_approx_matched[users_approx_matched.user_id.duplicated(keep=False)]
duplicated_users_in_ratebeer = users_approx_matched[users_approx_matched["user_id.1"].duplicated(keep=False)]
print(len(duplicated_users_in_advocate))
print(len(duplicated_users_in_ratebeer))
#Just verify that the one found earlier is here:
print(len(duplicated_users_in_advocate[duplicated_users_in_advocate.user_id=="magicuenca.1185749"]))
print(len(users_matched[users_matched.user_id == "magicuenca.1185749"])) #And so there is only one in the main file
#It is here, so again some users are duplicated

0
0
47
0
2
1


Again there are duplicates in the dataset. This time ratebeer has multiple users linked to a single users, but only in the approximation  not the main file. COntrary to the breweries the duplicates are in the ratebeer and not advocate. 

**BEERS DUPLICATES**

In [59]:
beers_matched = pd.read_csv(os.path.join(MATCHED_PATH,'beers.csv'),header=1)
beers_advocate = pd.read_csv(os.path.join(ADVOCATE_PATH,'beers.csv'))
beers_ratebeer = pd.read_csv(os.path.join(RATEBEER_PATH,'beers.csv'))

print(len(beers_matched))
print(len(beers_advocate))
print(len(beers_ratebeer))


45640
280823
442081


In [65]:
duplicated_beers_in_advocate = beers_matched[beers_matched.beer_id.duplicated(keep=False)]
duplicated_beers_in_ratebeer = beers_matched[beers_matched["beer_id.1"].duplicated(keep=False)]
print(len(duplicated_beers_in_advocate))
print(len(duplicated_beers_in_ratebeer))

0
0


In [66]:
beers_matched.beer_id.head(5)

0     19827
1     20841
2     20842
3     22659
4    178681
Name: beer_id, dtype: int64

In [67]:
beers_matched["beer_id.1"].head(5)

0     37923
1     41286
2     41287
3     41285
4    230283
Name: beer_id.1, dtype: int64

In [69]:
beers_ratebeer_solo = beers_ratebeer[~beers_ratebeer.beer_id.isin(beers_matched['beer_id.1'])]
beers_advocate_solo = beers_advocate[~beers_advocate.beer_id.isin(beers_matched['beer_id'])]

total_len = len(beers_ratebeer_solo) + len(beers_advocate_solo) + len(beers_matched)
print("Previous of the three datasets:\n-advocate:",len(beers_advocate),"\n-matched:", len(beers_matched),"\n-ratebeer:", len(beers_ratebeer))
print("New length of of the three datasets:\n-advocate:",len(beers_advocate_solo),"\n-matched:", len(beers_matched),"\n-ratebeer:", len(beers_ratebeer_solo))
print("The unique number of breweries is:", total_len)

Previous of the three datasets:
-advocate: 280823 
-matched: 45640 
-ratebeer: 442081
New length of of the three datasets:
-advocate: 235183 
-matched: 45640 
-ratebeer: 396441
The unique number of breweries is: 677264


No duplicates for the beers. Much seempler to treat XD (For duplication only :(   )

**Ratings analysis**

The ratings has multiple files for the different ratings/reviews.

In [132]:
ratings_matched = pd.read_csv(os.path.join(MATCHED_PATH,'ratings.csv'),header=1)
ratings_ba_matched = pd.read_csv(os.path.join(MATCHED_PATH,'ratings_ba.csv'))
ratings_ba_with_text_matched = pd.read_csv(os.path.join(MATCHED_PATH,'ratings_ba_with_text.csv'))
ratings_rb_matched = pd.read_csv(os.path.join(MATCHED_PATH,'ratings_rb.csv'))
ratings_rb_with_text_matched = pd.read_csv(os.path.join(MATCHED_PATH,'ratings_rb_with_text.csv'))

print(len(ratings_matched))
print(len(ratings_ba_matched))
print(len(ratings_ba_with_text_matched))
print(len(ratings_rb_matched))
print(len(ratings_rb_with_text_matched))


21964
955968
289025
1020638
1020599


The different files have different categories for the attributes (here we care about matched):
- ratings.csv: appearance, aroma, overall, palate, rating, review, style, taste, appearance.1, aroma.1, overall.1, palate.1, ratings.1, style.1, taste.1 and text.1
    review just tells us if there is a text or not => all the attributes are duplicated for both datasets
- ratings_ba.csv: style, appearance, aroma, palate, taste, overall rating, text, review
- ratings_ba_with_text.csv: style, appearance, aroma, palate, taste, overall rating, text, review => they have the same columns just thay make sure that the later one contains text.
- ratings_rb.csv: style, appearance, aroma, palate, taste, overall rating, text
- ratings_rb_with_text.csv: style, appearance, aroma, palate, taste, overall rating, text => they have the same columns just thay make sure that the later one contains text.
Only difference between the last four files if there is a review column.

In [77]:
print(ratings_matched.palate==ratings_matched["palate.1"])

0        False
1        False
2         True
3         True
4         True
         ...  
21959     True
21960     True
21961    False
21962    False
21963    False
Length: 21964, dtype: bool


We checked for the text and with palate, the attributes can differ. 

In [140]:
ratings_rb_with_text_matched.columns

Index(['beer_name', 'beer_id', 'brewery_name', 'brewery_id', 'style', 'abv',
       'date', 'user_name', 'user_id', 'appearance', 'aroma', 'palate',
       'taste', 'overall', 'rating', 'text'],
      dtype='object')

What makes ratings matched? Lets look at the date or beer id.

In [86]:
ratings_matched[["date","date.1"]].sample(5)

date      date.1
10181  1454497200  1394622000
3993   1419159600  1419159600
12531  1211018400  1211018400
6123   1467540000  1467453600
20155  1278237600  1278151200

In [128]:
N = 1000
beer_id_advocate = ratings_matched.iloc[N].beer_id
beer_id_ratebeer = ratings_matched.iloc[N]["beer_id.1"]

user_id_advocate = ratings_matched.iloc[N].user_id
user_id_ratebeer = ratings_matched.iloc[N]["user_id.1"]

beers_matched[beers_matched.beer_id == beer_id_advocate]==beers_matched[beers_matched["beer_id.1"] == beer_id_ratebeer]
users_matched[users_matched.user_id == user_id_advocate]==users_matched[users_matched["user_id.1"] == user_id_ratebeer]

joined  location  nbr_ratings  nbr_reviews  user_id  user_name  \
29    True      True         True         True     True       True   

    user_name_lower  joined.1  location.1  nbr_ratings.1  user_id.1  \
29             True      True        True           True       True   

    user_name.1  user_name_lower.1  
29         True               True

No its not the date of the comment, the beer drank makes a beer in match, if the beer is in both files (in the matched beer dataset). Now maybe one is not rated on the other dataset. Let us check that by looking at Nan's ie verify that a matched comment is a beer that has been commented on both websites.

In [129]:
advocate_criterions = ["appearance", "aroma", "overall", "palate", "rating", "review", "style", "taste"]
ratebeer_criterions = ["appearance.1", "aroma.1", "overall.1", "palate.1", "rating.1", "style.1", "taste.1"]

for i,attribute in enumerate(advocate_criterions+ratebeer_criterions):
    nan_ratings = ratings_matched[ratings_matched[attribute].isna()]
    print("Number of Nan values for ", attribute, ":", len(nan_ratings))


Number of Nan values for  appearance : 2669
Number of Nan values for  aroma : 2669
Number of Nan values for  overall : 2669
Number of Nan values for  palate : 2669
Number of Nan values for  rating : 0
Number of Nan values for  review : 0
Number of Nan values for  style : 0
Number of Nan values for  taste : 2669
Number of Nan values for  appearance.1 : 0
Number of Nan values for  aroma.1 : 0
Number of Nan values for  overall.1 : 0
Number of Nan values for  palate.1 : 0
Number of Nan values for  rating.1 : 0
Number of Nan values for  style.1 : 0
Number of Nan values for  taste.1 : 0


So from what wera seeing it is matches of comments made about a same beer, by the same user. He may have not answered one column.

Why are certain ratings in ratings_ba and ratings_rb. We suppose that it is because the user and the beer has been identify as a match. Now let's check this hypothesis.

In [185]:
N = 1200

user_id = ratings_ba_matched.iloc[N]["user_id"]
beer_id = ratings_ba_matched.iloc[N]["beer_id"]
print(user_id, beer_id)
print(len(beers_matched[beers_matched["beer_id"]==beer_id]))
print(len(users_matched[users_matched["user_id"]==user_id]))

dcmchew.760280 23358
1
1


In [188]:
N = 1200

user_id = ratings_rb_matched.iloc[N]["user_id"]
beer_id = ratings_rb_matched.iloc[N]["beer_id"]
print(user_id, beer_id)
print(len(beers_matched[beers_matched["beer_id.1"]==beer_id]))
print(len(users_matched[users_matched["user_id.1"]==user_id]))

17095 300727
1
1


Bingo it seems that the ratings in ba/rb matched corresponds to user matching that have made a comment about a beer that has a match. We checked it by hand for one value now lets check it for all ratings in ratings_ba and rb.

In [201]:
all_beers_matched = ratings_ba_matched["beer_id"].isin(beers_matched["beer_id"])
all_users_matched = ratings_ba_matched["user_id"].isin(users_matched["user_id"])
all_ratings_have_match_and = all_beers_matched & all_users_matched
all_ratings_have_match_or = all_beers_matched | all_users_matched

if all_ratings_have_match_and.all():
    print("All ratings are for a matching beer and user.")
if all_ratings_have_match_or.all():
    print("All ratings are for a matching beer or user.")
    print("There were {} with at least the beer or user having a match but {} had the user and beer match for a {} number of comment".format(all_ratings_have_match_or.sum(),all_ratings_have_match_and.sum(), len(all_ratings_have_match_and)))

All ratings are for a matching beer or user.
There were 955968 with at least the beer or user having a match but 78544 had the user and beer match for a 955968 number of comment


In [202]:
all_beers_matched = ratings_rb_matched["beer_id"].isin(beers_matched["beer_id.1"])
all_users_matched = ratings_rb_matched["user_id"].isin(users_matched["user_id.1"])
all_ratings_have_match_and = all_beers_matched & all_users_matched
all_ratings_have_match_or = all_beers_matched | all_users_matched

if all_ratings_have_match_and.all():
    print("All ratings are for a matching beer and user.")
if all_ratings_have_match_or.all():
    print("All ratings are for a matching beer or user.")
    print("There were {} with at least the beer or user having a match but {} had the user and beer match for a {} number of comment".format(all_ratings_have_match_or.sum(),all_ratings_have_match_and.sum(), len(all_ratings_have_match_and)))

All ratings are for a matching beer or user.
There were 1020638 with at least the beer or user having a match but 122372 had the user and beer match for a 1020638 number of comment


OH NOOOOO. It is not for a matching beer and user, but rather if either the beer or the user has a match.

Lets us check if one dataset contains more often a nan value or not:

In [147]:
for i,attribute in enumerate(ratings_ba_matched.columns):
    nan_ratings = ratings_ba_matched[ratings_ba_matched[attribute].isna()]
    print("Number of Nan values for ", attribute, ":", len(nan_ratings))

Number of Nan values for  beer_name : 0
Number of Nan values for  beer_id : 0
Number of Nan values for  brewery_name : 0
Number of Nan values for  brewery_id : 0
Number of Nan values for  style : 0
Number of Nan values for  abv : 0
Number of Nan values for  date : 0
Number of Nan values for  user_name : 5
Number of Nan values for  user_id : 0
Number of Nan values for  appearance : 363750
Number of Nan values for  aroma : 363750
Number of Nan values for  palate : 363750
Number of Nan values for  taste : 363750
Number of Nan values for  overall : 363750
Number of Nan values for  rating : 0
Number of Nan values for  text : 666943
Number of Nan values for  review : 0


In [ ]:
for i,attribute in enumerate(ratings_rb_matched.columns):
    nan_ratings = ratings_rb_matched[ratings_rb_matched[attribute].isna()]
    print("Number of Nan values for ", attribute, ":", len(nan_ratings))

Number of Nan values for  beer_name : 0
Number of Nan values for  beer_id : 0
Number of Nan values for  brewery_name : 0
Number of Nan values for  brewery_id : 0
Number of Nan values for  style : 0
Number of Nan values for  abv : 0
Number of Nan values for  date : 0
Number of Nan values for  user_name : 0
Number of Nan values for  user_id : 0
Number of Nan values for  appearance : 0
Number of Nan values for  aroma : 0
Number of Nan values for  palate : 0
Number of Nan values for  taste : 0
Number of Nan values for  overall : 0
Number of Nan values for  rating : 0
Number of Nan values for  text : 39


Advocates match ratings contains certain attribute not rated by times. For ratebeer this is never the case. In both cases a text may be missing.

In [173]:
ratings_rb_matched.sample(5)

beer_name  beer_id  \
475045               Magnify Seasonally Appropriate   357019   
361213                       Furthermore Knot Stock    61941   
79399                                Malthus Baluba    44669   
707230  Jolly Pumpkin Fuego del Otono (Autumn Fire)    53557   
351090                              Kinn Gamleguten   207987   

                      brewery_name  brewery_id                 style  abv  \
475045     Magnify Brewing Company       23248       Berliner Weisse  4.5   
361213                 Furthermore        6951  Spice/Herb/Vegetable  5.5   
79399        Il Birrificio di Como        5735  Spice/Herb/Vegetable  6.9   
707230  Jolly Pumpkin Artisan Ales        4923           Belgian Ale  6.1   
351090               Kinn Bryggeri       11098               Old Ale  7.0   

              date     user_name  user_id  appearance  aroma  palate  taste  \
475045  1441188000        vtafro    85502           5      5       5      7   
361213  1251108000  zizzybalubba    57664           3      6       3      4   
79399   1137063600     rauchbier     2463           3      8       3      8   
707230  1370944800        OKplus   237085           3      6       3      8   
351090  1401098400       fonefan    38926           3      6       3      7   

        overall  rating                                               text  
475045       13     3.5  Draft. Golden pour with white lacing. Light fr...  
361213        9     2.5  12 oz. bottle from Steve’s Liquors in Madiso...  
79399        15     3.7  Bottle, courtesy of Lorenzo @ ZBF 2005. Deep c...  
707230       16     3.6  Split this one with the wife.750ml served into...  
351090       13     3.2  Bottle 750ml. @ [ Fastelavn Smagning - Carniva...